### DISCUSSION TITLE
Simulating thousands of nodes

### DISCUSSION THREAD

**Discussion by haxxpop** [shadow/shadow#2622] [Q&A] [answered]
I'm trying to simulate the network of 5,000 nodes of [gossipsub protocol](https://docs.libp2p.io/concepts/pubsub/overview/)

Since running a lot processes requires a lot of memory, I created some swap file of 230G as described [here](https://www.digitalocean.com/community/tutorials/how-to-add-swap-space-on-ubuntu-22-04)

However, it seems not easy to just use the swap file as the memory. As you can see in the figure, I have 230G of swap space, but it's not used at all. As a result, I can run only 400 nodes (with 2GB of memory).

**Does anyone have experience running thousands of nodes?** I think even though each process uses very little memory, in the end the total sum of memory will be very high if you simulate a lot of nodes.

What I'm guessing now is that most memory used by each process is shared memory and the shared memory probably cannot be swapped out?

The following is the htop result. Let me know if you need any clarification or some source code.

![Screenshot from 2022-12-20 17-20-14](https://user-images.githubusercontent.com/631508/208654315-64b41905-3cc8-4b40-a30a-1070c859193e.png)

**Comment by rwails** [comment, marked as answer]
Thanks for your question!

> Does anyone have experience running thousands of nodes?

Yes! Shadow is routinely used for simulations with many thousands of nodes. However, it's usually done on machines with many more resources than the one you've posted.

> and the shared memory probably cannot be swapped out?

I do believe that Linux kernel memory can't be paged out [0], and simulating many processes can consume a decent amount of memory in the kernel per process (a `task_struct` kernel object alone might consume 1-2 kilobytes per process). Also, you might need to adjust your machine's virtual memory parameters before the right swapping behavior is achieved [1].

> I created some swap file of 230G

To the broader question, I am worried about the strategy of running a large experiment and trying to fit all the state into swap space. Even if you worked out the paging, the experiment would likely progress too slowly to be useful. There are a couple of ways you might consider adjusting your process:

1. The lowest effort would be to run multiple simulations using a smaller number of nodes (say, 50) and try to extrapolate the results to a larger number of nodes. Details about this strategy can be found in [2].

2. Write custom software that simulates the behavior/activity of multiple protocol clients. That way many gossipsub protocol nodes can be run using fewer Linux processes.

3. The most costly, most obvious alternative would be to find time on a larger machine. Shadow can be run via Docker, and so running an experiment hosted on the cloud shouldn't be too painful to set up.

[0] <https://tldp.org/LDP/khg/HyperNews/get/memory/linuxmm.html>
[1] <https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/6/html/performance_tuning_guide/s-memory-tunables>
[2] <https://www.robgjansen.com/publications/neverenough-sec2021.pdf>

**Reply by sporksmith** [reply]
+1 to all of that, but one small thing you could try is disabling the memory manager, which otherwise maps all of the managed process heaps and stacks into shared memory for faster IPC, and yeah I'm not sure whether that memory is eligible to be swapped.

`--use-memory-manager=false` https://shadow.github.io/docs/guide/shadow_config_spec.html#experimentaluse_memory_manager

**Reply by haxxpop** [reply]
I tried using `--use-memory-manager=false` already. It works for me. I can run 5,000 gossipsub nodes with an 8GB machine with 96GB of swap space used. Thank you.

![Screenshot from 2022-12-23 14-27-16](https://user-images.githubusercontent.com/631508/209319089-407018b7-5f53-4c70-a9c0-2cab2c674f06.png)

**Reply by sporksmith** [reply]
Glad to hear it!

I think `--use-memory-manager=false` is probably the right solution for your scenario, but for completeness:

One thing I forgot is that you'd probably need to remount `/dev/shm` with a larger size for `--use-memory-manager=true` (the default) to work. This is where the memory manager gets its memory for remapping. Most systems only configure it to be half the size of physical RAM. AFAIK you *could* configure it to use > 100% of physical RAM, s.t. it ends up using swap. e.g. see https://www.cyberciti.biz/tips/what-is-devshm-and-its-practical-usage.html

Since you already have something working it's *probably* not worth messing with unless you're curious or it's worth your time to slightly improve simulation performance. It used to be a bigger win, but we've improved the *non* memory-mapped code path to the point that it's only about a ~1% performance gain https://github.com/shadow/shadow/issues/2581
